### PR TITLE
Feature/permute vertex

### PR DESCRIPTION
**Pull request by GeorgeNash**
Permute.

Can be used to transpose and more.

### CODE REVIEW

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertex.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    @Override
+    public Map<Vertex, PartialDerivative> reverseModeAutoDifferentiation(PartialDerivative derivativeOfOutputWithRespectToSelf) {
+        Map<Vertex, PartialDerivative> partials = new HashMap<>();
+        int[] permuteToApply = reversePermute(derivativeOfOutputWithRespectToSelf);
+        partials.put(inputVertex, derivativeOfOutputWithRespectToSelf.permute(permuteToApply));
+        return partials;
+    }
+
+    private int[] forwardPermute(PartialDerivative partial) {
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertex.java]: No need to iterate twice nor do you need the wrtShape.
```
    private int[] forwardPermute(PartialDerivative partial) {

        int[] permuteToApply = new int[partial.get().getRank()];

        for (int i = 0; i < partial.get().getRank(); i++) {
            if (i < rearrange.length) {
                permuteToApply[i] = rearrange[i];
            } else {
                permuteToApply[i] = i;
            }
        }

        return permuteToApply;
    }
```

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertex.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        }
+
+        for (int j = 0; j < wrtShape.length; j++) {
+            permuteToApply[j + rearrange.length] = j + rearrange.length;
+        }
+
+        return permuteToApply;
+    }
+
+    private int[] reversePermute(PartialDerivative partial) {
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertex.java]: The n^2 for loop is nasty and not needed. You can calculate what the inversion of the permute will be at construction time using something like:

```
...
   this.invertedRearrange = invertedPermute(rearrange);
...
    private static int[] invertedPermute(int[] rearrange) {
        int[] inverted = new int[rearrange.length];

        for (int i = 0; i < rearrange.length; i++) {
            inverted[rearrange[i]] = i;
        }

        return inverted;
    }
```
And then later just adjust it for the rank of the incoming partial:
```
    private int[] reversePermute(PartialDerivative partial) {

        int partialRank = partial.get().getRank();
        int[] permuteToApply = new int[partialRank];
        int ofRank = partialRank - this.getRank();

        for (int i = 0; i < partialRank; i++) {
            if (i >= ofRank) {
                permuteToApply[i] = invertedRearrange[i - ofRank] + ofRank;
            } else {
                permuteToApply[i] = i;
            }
        }

        return permuteToApply;
    }
```

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertexTest.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        a.setValue(DoubleTensor.create(new double[]{1, 2, 3, 4, 5, 6}, 2, 3));
+
+        PermuteVertex transpose = new PermuteVertex(a, 1, 0);
+
+        Assert.assertArrayEquals(new long[]{3, 2}, transpose.getShape());
+        Assert.assertArrayEquals(a.getValue().transpose().asFlatDoubleArray(), transpose.getValue().asFlatDoubleArray(), 1e-6);
+    }
+
+    @Test
+    public void changesMatchGradientRankTwo() {
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertexTest.java]: Test a few more permute options...
```
    @Test
    public void changesMatchGradientRankTwo() {
        testWithFiniteDifference(v -> v.permute(1, 0), new long[]{4, 5});
        testWithFiniteDifference(v -> v.permute(0, 1), new long[]{4, 5});
    }

    @Test
    public void changesMatchGradientRankThree() {
        testWithFiniteDifference(v -> v.permute(1, 2, 0), new long[]{4, 3, 2});
        testWithFiniteDifference(v -> v.permute(0, 1, 2), new long[]{4, 3, 2});
        testWithFiniteDifference(v -> v.permute(2, 1, 0), new long[]{4, 3, 2});
    }
```

File: build.gradle [hunk 4]
```diff
@@ -23,7 +23,7 @@ allprojects {
 }
 
 dependencies {
-    compile project(":keanu-project")
+    compile project(':keanu-project')
```
[the remark this answers is not in the document]
Comment (GeorgeNash) [reply build.gradle]: Not too sure - my formatter...

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivative.java [hunk 5]
```diff
@@ -123,6 +123,11 @@ public PartialDerivative multiplyAlongWrtDimensions(DoubleTensor multiplier) {
         return new PartialDerivative(result);
     }
 
+    public PartialDerivative permute(int[] rearrange) {
```
[the remark this answers is not in the document]
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivative.java]: Initially I was calculating the permute to apply inside there. It's a bit pointless now that I calculate those in separate methods, I'll take it out

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertex.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                    reversedPermute[j + ofShape.length] = k + ofShape.length;
+                }
+            }
+        }
+
+        return reversedPermute;
+    }
+
+    private static long[] calculatePermutedShape(DoubleVertex inputVertex, int... rearrange) {
+        DoubleTensor inputShape = DoubleTensor.create(Arrays.stream(inputVertex.getShape()).asDoubleStream().toArray(), inputVertex.getShape());
```
[the remark this answers is not in the document]
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertex.java]: That's what I'm doing, it needs to be a tensor to apply the rearrange

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertex.java (hunk 1 above)
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertex.java]: I thought splitting it into two smaller iterations would make it clearer. Yours does look a lot clearer given we don't need all of my `ofShape` and `wrtShape`!

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertex.java (hunk 2 above)
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertex.java]: nice!

File: docs/release_notes.txt [hunk 7]
```diff
@@ -1,3 +0,0 @@
-### Version 0.0.19 ###
```
Comment (gordoncaleb) [inline docs/release_notes.txt]: Why deleted?

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertexTest.java [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+        UniformVertex B = new UniformVertex(0, 10);
+        B.setValue(DoubleTensor.arange(0, 6).reshape(1, 2, 3));
+
+        MultiplicationVertex C = A.times(B);
+
+        PermuteVertex permute = C.permute(2, 0, 1);
+        PermuteVertex revertThePermute = permute.permute(1, 2, 0);
+
+        Assert.assertArrayEquals(C.getValue().asFlatDoubleArray(), revertThePermute.getValue().asFlatDoubleArray(), 1e-6);
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertexTest.java]: Comparing only the flat arrays doesn't check for shape issues. You can use one of Mike's hamcrest matchers or assert that the two tensors are equal.

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertexTest.java [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+
+    @Test
+    public void changesMatchGradientRankThree() {
+        testWithFiniteDifference(v -> v.permute(1, 2, 0), new long[]{4, 3, 2});
+        testWithFiniteDifference(v -> v.permute(0, 1, 2), new long[]{4, 3, 2});
+        testWithFiniteDifference(v -> v.permute(2, 1, 0), new long[]{4, 3, 2});
+    }
+
+    @Test
+    public void canCalculateAutoDiffOfRankThreePermute() {
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertexTest.java]: A lot of these tests aren't testing much more than what the finite difference tests are testing. What more they're testing is really more of a Differentiator or other operation (e.g. times, sum) concern.

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertexTest.java (hunk 8 above)
Comment (GeorgeNash) [reply keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/PermuteVertexTest.java]: Changed to be a matcher

**Pull request merged**